src/server/worker/pipeline/transform/status.go

```diff
@@ -98,3 +102,41 @@ func (s *Status) Cancel(jobID string, datumFilter []string) bool {
 	}
 	return false
 }
+
+func (s *Status) withDatumBatch(cb func(<-chan error, chan<- []string) error) error {
+	s.withBatchLock(func() {
+		if s.nextChan != nil {
+			panic("multiple goroutines attempting to set up a datum batch")
+		}
+		s.nextChan, s.setupChan = make(chan error), make(chan []string)
+	})
+	defer s.withBatchLock(func() {
+		s.nextChan, s.setupChan = nil, nil
+	})
+	return cb(s.nextChan, s.setupChan)
+}
+
+func (s *Status) withBatchLock(cb func()) {
+	s.batchMutex.Lock()
+	cb()
+	s.batchMutex.Unlock()
```

Comment:
I think it's better practice to defer here, because that unlocks during a panic.  Not that it really matters for us.

Reply:
yeah, I chose this for consistency with the other with lock method. I went ahead and changed both to the defer unlock pattern.